include/itkVkMultiResolutionPyramidImageFilter.hxx

```diff
@@ -0,0 +1,210 @@
+
+  using ImageToImageType = ImageToImageFilter<TOutputImage, TOutputImage>;
+  using ResampleShrinkerType = ResampleImageFilter<TOutputImage, TOutputImage>;
+  using ShrinkerType = ShrinkImageFilter<TOutputImage, TOutputImage>;
+
+  auto                               caster = CasterType::New();
+  auto                               spatialSmoother = SpatialSmootherType::New();
+  auto                               fftSmoother = FFTSmootherType::New();
+  typename BaseSmootherType::Pointer smoother;
+
+  typename ImageToImageType::Pointer shrinkerFilter;
+  //
+  // only one of these pointers is going to be valid, depending on the
+  // value of UseShrinkImageFilter flag
+  typename ResampleShrinkerType::Pointer resampleShrinker;
+  typename ShrinkerType::Pointer         shrinker;
+
+  if (this->GetUseShrinkImageFilter())
+  {
+    shrinker = ShrinkerType::New();
+    shrinkerFilter = shrinker.GetPointer();
+  }
+  else
+  {
+    resampleShrinker = ResampleShrinkerType::New();
+    using LinearInterpolatorType = itk::LinearInterpolateImageFunction<OutputImageType, double>;
+    auto interpolator = LinearInterpolatorType::New();
+    resampleShrinker->SetInterpolator(interpolator);
+    resampleShrinker->SetDefaultPixelValue(0);
+    shrinkerFilter = resampleShrinker.GetPointer();
+  }
+  // Setup the filters
+  caster->SetInput(inputPtr);
+
+  // Set up smoothing filters
+  spatialSmoother->SetUseImageSpacing(false);
+  spatialSmoother->SetInput(caster->GetOutput());
+  spatialSmoother->SetMaximumError(this->m_MaximumError);
+  fftSmoother->SetUseImageSpacing(false);
+  fftSmoother->SetInput(caster->GetOutput());
+  fftSmoother->SetMaximumError(this->m_MaximumError);
+
+  unsigned int ilevel, idim;
+  unsigned int factors[ImageDimension];
+  VarianceType variance;
+
+  for (ilevel = 0; ilevel < this->m_NumberOfLevels; ++ilevel)
+  {
+    this->UpdateProgress(static_cast<float>(ilevel) / static_cast<float>(this->m_NumberOfLevels));
+
+    // Allocate memory for each output
+    OutputImagePointer outputPtr = this->GetOutput(ilevel);
+    outputPtr->SetBufferedRegion(outputPtr->GetRequestedRegion());
+    outputPtr->Allocate();
+
+    // compute shrink factors
+    for (idim = 0; idim < ImageDimension; ++idim)
+    {
+      factors[idim] = this->m_Schedule[ilevel][idim];
+    }
+
+    if (!this->GetUseShrinkImageFilter())
+    {
+      using IdentityTransformType = itk::IdentityTransform<double, OutputImageType::ImageDimension>;
+      auto identityTransform = IdentityTransformType::New();
+      resampleShrinker->SetOutputParametersFromImage(outputPtr);
+      resampleShrinker->SetTransform(identityTransform);
+    }
+    else
+    {
+      shrinker->SetShrinkFactors(factors);
+    }
+
+    // select spatial or FFT smoothing based on user threshold settings
+    // to maximize anticipated performance
+    if (GetUseFFT(this->GetKernelRadius(ilevel)))
+    {
+      smoother = fftSmoother;
```

Comment:
This seems to work as is.  I suspect the underlying implementation is essentially
```cpp
smoother = static_cast<BaseSmootherType *>(fftSmoother.GetPointer());
```
Perhaps it would make sense to show that explicitly?  Likewise a few lines below, even though `spatialSmoother` and `smoother` happen to be the same type.

Reply:
Added.